src/ORM/DataObject.php

```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
         }
         return $this;
     }
+
+    /**
+     * Find objects in the given relationships, merging them into the given list
+     *
+     * @param string $source Config property to extract relationships from
+     * @param bool $recursive True if recursive
+     * @param ArrayList $list Optional list to add items to
+     * @return ArrayList The list
+     */
+    public function findRelatedObjects($source, $recursive = true, $list = null)
+    {
+        if (!$list) {
+            $list = new ArrayList();
+        }
+
+        // Skip search for unsaved records
+        if (!$this->isInDB()) {
+            return $list;
+        }
+
+        $relationships = $this->config()->get($source) ?: [];
+        foreach ($relationships as $relationship) {
+            // Warn if invalid config
+            if (!$this->hasMethod($relationship)) {
+                trigger_error(sprintf(
+                    "Invalid %s config value \"%s\" on object on class \"%s\"",
+                    $source,
+                    $relationship,
+                    get_class($this)
+                ), E_USER_WARNING);
+                continue;
+            }
+
+            // Inspect value of this relationship
+            $items = $this->{$relationship}();
+
+            // Merge any new item
+            $newItems = $this->mergeRelatedObjects($list, $items);
+
+            // Recurse if necessary
+            if ($recursive) {
+                foreach ($newItems as $item) {
+                    /** @var DataObject $item */
+                    $item->findRelatedObjects($source, true, $list);
+                }
+            }
+        }
+        return $list;
+    }
+
+    /**
+     * Helper method to merge owned/owning items into a list.
+     * Items already present in the list will be skipped.
+     *
+     * @param ArrayList $list Items to merge into
+     * @param mixed $items List of new items to merge
+     * @return ArrayList List of all newly added items that did not already exist in $list
+     */
+    public function mergeRelatedObjects($list, $items)
+    {
+        $added = new ArrayList();
+        if (!$items) {
+            return $added;
+        }
+        if ($items instanceof DataObject) {
+            $items = [$items];
+        }
+
+        /** @var DataObject $item */
+        foreach ($items as $item) {
+            $this->mergeRelatedObject($list, $added, $item);
+        }
+        return $added;
+    }
+
+    /**
+     * Merge single object into a list
+     *
+     * @param ArrayList $list Global list. Object will not be added if already added to this list.
+     * @param ArrayList $added Additional list to insert into
+     * @param DataObject $item Item to add
+     */
+    public function mergeRelatedObject($list, $added, $item)
```

Comment:
Fair enough, I'm probably too critical about what's public, only half way through the Versioned pull request